src/NServiceBus.Gateway/Recoverability/RetryFailedGatewayMessagesBehavior.cs

```diff
@@ -0,0 +1,97 @@
+﻿namespace NServiceBus.Gateway
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using NServiceBus.DelayedDelivery;
+    using NServiceBus.DeliveryConstraints;
+    using NServiceBus.Logging;
+    using NServiceBus.Pipeline;
+    using NServiceBus.Transports;
+
+    class RetryFailedGatewayMessagesBehavior : ForkConnector<ITransportReceiveContext, IRoutingContext>
+    {
+        public RetryFailedGatewayMessagesBehavior(string localAddress, Func<IncomingMessage, Exception, int, TimeSpan> retryPolicy)
+        {
+            this.localAddress = localAddress;
+            this.retryPolicy = retryPolicy;
+        }
+
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next, Func<IRoutingContext, Task> fork)
+        {
+            try
+            {
+                await next().ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                var sentForRetry = await SendForRetry(context, exception, fork).ConfigureAwait(false);
+                if (!sentForRetry)
+                {
+                    throw;
+                }
+            }
+        }
+
+        private async Task<bool> SendForRetry(ITransportReceiveContext context, Exception exception, Func<IRoutingContext, Task> fork)
+        {
+            var message = context.Message;
+
+            var currentRetry = GetCurrentRetry(message);
+
+            var timeIncrease = retryPolicy(message, exception, currentRetry);
+
+            if (timeIncrease <= TimeSpan.Zero)
+            {
+                return false;
+            }
+
+            Logger.Warn($"Gateway message '{message.MessageId}' failed. Will reschedule message after {timeIncrease}:", exception);
+
+            var messageToRetry = new OutgoingMessage(message.MessageId, message.Headers, message.Body);
+
+            messageToRetry.Headers[Headers.Retries] = currentRetry.ToString();
+            messageToRetry.Headers[Headers.RetriesTimestamp] = DateTimeExtensions.ToWireFormattedString(DateTime.UtcNow);
+
+            var dispatchContext = this.CreateRoutingContext(messageToRetry, localAddress, context);
+
+            context.Extensions.Set(new List<DeliveryConstraint>
+            {
+                new DelayDeliveryWith(timeIncrease)
+            });
+
+            await fork(dispatchContext).ConfigureAwait(false);
+
+            return true;
+        }
+
+        static int GetCurrentRetry(IncomingMessage message)
+        {
+            string value;
+            if (message.Headers.TryGetValue(Headers.Retries, out value))
+            {
+                int i;
+                if (int.TryParse(value, out i))
+                {
+                    return i + 1;
+                }
+            }
+            return 1;
```

Comment:
this seems weird and like an attempt to reimplement the configuration behavior from NSB 5 (e.g. MaxRetries=2 == 1 retry attempt). If there was no retry, returning 0 would make more sense to me?